File: py/ch02/E02C_profile_most_probable_kmer.py
# Profile-most Probable k-mer Problem:
# Find a Profile-most probable k-mer in a string
#   Input: A string Text, an integer k, and a 4 x k matrix Profile
#   Output: A Profile-most probable k-mer in Text

import pandas as pd

from utils import text, generate_probs, bases


def MostProbableKmer(genome, k, matrix=None):
    if matrix is None:
        matrix = generate_probs(k)
    else:
        matrix = pd.DataFrame(matrix).T
        matrix.columns = bases

    probs = []
    max_prob = -1
    max_kmer = ""

    for i in range(len(genome) - k):
        substr = text(genome, i, k)
        total = 1

        for j in range(len(substr)):
            total *= matrix.loc[j, substr[j]]

        probs.append(total)

        if total > max_prob:
            max_prob = total
            max_kmer = substr

    return max_kmer


if __name__ == "__main__":
    _genome = input("Genome: ").upper()
    _k = int(input("k: "))

    print(MostProbableKmer(_genome, _k))

File: cpp/ch02/E02C_profile_most_probable_kmer.cpp
//
// Created by Matej Hrnčiar on 28/07/2022.
//

#include "E02C_profile_most_probable_kmer.h"

string MostProbableKmer(const string &genome, int k, vector< vector<float> > matrix) {
    if (matrix.empty()) {
        matrix = GenerateProbs(k);
    }

    list<float> probs;
    float max_prob = -1;
    string max_kmer;

    for (int i = 0; i < genome.size() - k; i++) {
        string substr = Text(genome, i, k);
        float total = 1;

        for (int j = 0; j < substr.size(); j++) {
            total *= matrix[symbol_key.find(substr[j])->second][j];
        }

        probs.push_back(total);

        if (total > max_prob) {
            max_prob = total;
            max_kmer = substr;
        }
    }

    return max_kmer;
}

/*
int main() {
    string _genome;
    int _k;
    vector< vector<float> > matrix;

    cout << "Genome: ";
    cin >> _genome;
    cout << "k: ";
    cin >> _k;

    cout << MostProbableKmer(_genome, _k, matrix);
}
*/
